### ISSUE TITLE
Text field inside of dynamic loses focus when re-rendered

### ISSUE THREAD

**Issue by lilactown** [HumbleUI/HumbleUI#50]
Currently, if you use a `text-field` inside of a `dynamic` component, the text-field will lose focus each time it is newly created.

A minimal example:

```clojure
(def *counter (atom 0))
(def *text (atom {:text ""}))

(reset! state/*app
        (ui/default-theme
         {}
         (ui/dynamic
          ctx
          [counter @*counter]
          (ui/column
           (ui/label counter)
           (ui/text-field *text)))))

(defn timer []
  (future
   (Thread/sleep 1000)
   (swap! *counter inc)
   (timer)))

(timer)
```

**Comment by lilactown** [comment]
The example above might be a bit contrived, as I could move the `dynamic` component to only contain the `(ui/label counter)`.

However, in a project I have my own implementation of `text-field` that I've forked and modified to have "disabled" and "error" states, along with button and in the future more components. The approach I'm taking is to use context for these, e.g.

```clojure
(def form
  (ui/dynamic
   ctx
   [disabled? (:disabled? @*form-state)]
   (disable disabled?
     (ui/column
      (text-field *form-state)
      (button on-submit (ui/label "Submit"))))))
```

This is very convenient, as I can group components together that should be disabled or have error borders & messages. However, because `text-field` now needs to be re-rendered by the `dynamic` parent with the new context, it loses its focus state.

**Comment by tonsky** [comment]
Yeah, that’s a known problem :( Dynamic basically re-creates entire subtree when its inputs change. New components lose parts of their state (even though everything about text-field is stored in atom, I guess I missed focus somehow). I plan to eventually replace dynamic with something more reliable, and also make sure that replacing components with similar ones doesn’t destroy any important state

**Comment by Quezion** [comment]
Tested on latest HumbleUI & can confirm this bug is still active. I was hoping some of the latest `main` commits about signals & vdom might've fixed it, but I suppose neither of those solve the component-state storage problem.

I'll continue using my crudely patched version from #71 for now. I'd be willing to help on this issue but I'm not sure of the next steps or vision for it.

**Comment by tonsky** [comment]
They could solve it, but the commits were just tests, not merged into main impl. I haven’t made a decision to go with them yet
